### DISCUSSION TITLE
Postmessage based live-preview

### DISCUSSION THREAD

**Discussion by br41nslug** [directus/directus#21695] [Feature Requests]
### Summary

The "Live Preview" is not so much live currently as it does not provide updated values without saving the record. There are plans for a [WebSocket based live-preview](https://github.com/directus/directus/discussions/18792) however I think a `window.postMessage` based approach could be both more reliable and easier to implement for end users like demonstrated here https://github.com/br41nslug/directus-extension-live-preview-sync

### Basic Example

Using the following snippet on the preview page the end user can start receiving events:
```ts
window.addEventListener("message", (event) => {
    const { type, path, values } = event.data;
    if (type === 'directus-preview') {
        // handle data update
        updatePage(event.data);
    }
}, false);
```

### Motivation

I believe this to be a more resilient and easy to set up solution for pushing updates to the live preview pane. Where for a WebSocket based implementation you would need the end user to connect to the socket, authenticate, subscribe, handle disconnects on the preview page and configure the backend to properly route persistent WebSockets.
While with this `postMessage` approach the data is pushed from the Directus App directly to the preview page in the browser without needing to connect over the network from either side.

### Detailed Design

Instead of having to add a new hidden interface to push these `window.postMessage` events Directus could directly include the logic into the item detail page which would allow the event to still be sent if the iframe is popped out into a separate window. Once the `iframe` is loaded an initial message can be pushed after which the logic can listen to the form data changes using 
```const values = inject('values');``` 

### Must Have:
- An on/off toggle for this feature

### Could Have:
- A way of specifying the structure of the initial data sent to the frame
- A way of specifying/limiting the structure of update data sent to the frame

### Won't Have:
- Unicorns

### Drawbacks

This approach is one way communication, the Directus App pushes the updates it sees locally 1 on 1, without an option for the preview page to specify what data it wants or fetch extra information.
A security drawback of this approach is that it does not check permissions beyond what the currently logged in user is allowed to see. This combined with a preview page you're not in control of or that is loading malicious 3rd party scripts can be used to extract/leak information that should have been private.

### Alternatives

WebSocket based live-preview: https://github.com/directus/directus/discussions/18792

### Adoption Strategy

It is a new feature, no real migration needed.

### Unresolved Questions

- extra or initial information to be passed to the preview?

**Comment by pavelstianko** [comment]
That sounds amazing! ❤️

**Comment by angelod1as** [comment]
Was this moved to the new feature request website? Is this planned? It would be great.
